From 8ec446cae206308f473bada3c93c0faf08771b66 Mon Sep 17 00:00:00 2001
From: haiqiang yao <haiqiangx.yao@intel.com>
Date: Thu, 12 Nov 2015 11:00:25 +0800
Subject: [PATCH 17/28] atomisp2:Add the interface to select isp pipe version

the patch adds a new IOCTL V4L2_CID_ATOMISP_SELECT_ISP_VERSION.
and Camera HAL select the ISP capture pipe version.currently only
2.2 and 2.7 are avaliable.

Change-Id: I6851595d476159129ab869fd1d7b99fde93bdc94
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-9279
Signed-off-by: haiqiang yao <haiqiangx.yao@intel.com>
Original-Tracked-On: #H1504088527
Reviewed-on: https://icggerrit.ir.intel.com/84599
Build: Dessert M BXT RVP (SC)
Build: Dessert M COHO (SC)
Approver: Bin B Han <bin.b.han@intel.com>
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Signed-off-by: Yang Fan F <fan.f.yang@intel.com>
Reviewed-on: https://android.intel.com:443/440503
---
 .../pci/atomisp2/atomisp_driver/atomisp_common.h   |  2 ++
 .../atomisp2/atomisp_driver/atomisp_compat_css20.c | 10 ++++++++++
 .../pci/atomisp2/atomisp_driver/atomisp_subdev.c   | 22 ++++++++++++++++++++++
 .../pci/atomisp2/atomisp_driver/atomisp_subdev.h   |  1 +
 4 files changed, 35 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_common.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_common.h
index 91144f5..8cb8ea7 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_common.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_common.h
@@ -43,6 +43,8 @@ extern int pad_h;
 #define CSS_DTRACE_VERBOSITY_LEVEL	5	/* Controls trace verbosity */
 #define CSS_DTRACE_VERBOSITY_TIMEOUT	9	/* Verbosity on ISP timeout */
 #define MRFLD_MAX_ZOOM_FACTOR	1024
+#define ATOMISP_CSS_ISP_PIPE_VERSION_2_2    0
+#define ATOMISP_CSS_ISP_PIPE_VERSION_2_7    1
 
 #define IS_ISP2401(isp)							\
 	(((isp)->media_dev.hw_revision & ATOMISP_HW_REVISION_MASK)	\
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
index 445c45d..6c21277 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
@@ -718,6 +718,16 @@ static void __apply_additional_pipe_config(
 		if (stream_env->pipe_configs[pipe_id].
 			default_capture_config.mode == CSS_CAPTURE_MODE_RAW)
 			stream_env->pipe_configs[pipe_id].enable_dz = false;
+
+		/* the isp default to use ISP2.2 and the camera hal will
+		 * control whether use isp2.7 */
+		if (asd->select_isp_version->val ==
+			ATOMISP_CSS_ISP_PIPE_VERSION_2_7)
+			stream_env->pipe_configs[pipe_id].isp_pipe_version =
+				SH_CSS_ISP_PIPE_VERSION_2_7;
+		else
+			stream_env->pipe_configs[pipe_id].isp_pipe_version =
+				SH_CSS_ISP_PIPE_VERSION_2_2;
 		break;
 	case IA_CSS_PIPE_ID_VIDEO:
 		/* enable reduced pipe to have binary
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.c
index e67f6c6..63fce53 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.c
@@ -1033,6 +1033,24 @@ static const struct v4l2_ctrl_config ctrl_depth_mode = {
 	.def = 0,
 };
 
+/*
+ * Control for selectting ISP version
+ *
+ * When enabled, that means ISP version will be used ISP2.7. when disable, the
+ * isp will default to use ISP2.2.
+ * Note: Make sure set this configuration before creating stream.
+ */
+static const struct v4l2_ctrl_config ctrl_select_isp_version = {
+	.ops = &ctrl_ops,
+	.id = V4L2_CID_ATOMISP_SELECT_ISP_VERSION,
+	.type = V4L2_CTRL_TYPE_BOOLEAN,
+	.name = "Select Isp version",
+	.min = 0,
+	.max = 1,
+	.step = 1,
+	.def = 0,
+};
+
 #ifdef CONFIG_ION
 /*
  * Control for ISP ion device fd
@@ -1224,6 +1242,10 @@ static int isp_subdev_init_entities(struct atomisp_sub_device *asd)
 			v4l2_ctrl_new_custom(&asd->ctrl_handler,
 					     &ctrl_disable_dz,
 					     NULL);
+	asd->select_isp_version =
+			v4l2_ctrl_new_custom(&asd->ctrl_handler,
+					     &ctrl_select_isp_version,
+					     NULL);
 
 #ifdef CONFIG_ION
 	asd->ion_dev_fd =
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.h
index 5bbd49e..80e533b 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_subdev.h
@@ -318,6 +318,7 @@ struct atomisp_sub_device {
 	struct v4l2_ctrl *enable_raw_buffer_lock;
 	struct v4l2_ctrl *ion_dev_fd;
 	struct v4l2_ctrl *disable_dz;
+	struct v4l2_ctrl *select_isp_version;
 
 	struct {
 		struct list_head fw;
-- 
1.9.1

